daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/TypeSubst.scala

```diff
@@ -8,192 +8,55 @@ import com.digitalasset.daml.lf.language.Ast._
 import com.digitalasset.daml.lf.validation.Util._
 import com.digitalasset.daml.lf.validation.traversable.TypeTraversable
 
-private[validation] case class TypeSubst(map: Map[TypeVarName, Type], private val depth: Int = 0) {
-
-  lazy val freeVars: Set[TypeVarName] =
-    (Set.empty[TypeVarName] /: map.values)(TypeSubst.freeVars)
-
-  def apply(typ: Type): Type = typ match {
-    case TVar(name) => map.getOrElse(name, typ)
-    case TTyCon(_) | TBuiltin(_) | TNat(_) => typ
-    case TApp(t1, t2) => TApp(apply(t1), apply(t2))
-    case TForall((v, k), t) =>
-      val (v1, subst1) = if (freeVars.contains(v)) {
-        val v1 = freshTypeVarName
-        v1 -> TypeSubst(map + (v -> TVar(v1)))
-      } else
-        v -> TypeSubst(map - v)
-      TForall((v1, k), subst1(t))
-    case TStruct(ts) =>
-      TStruct(ts.transform { (_, x) =>
-        apply(x)
-      })
-  }
-
-  private def freshTypeVarName: TypeVarName =
+private[validation] object TypeSubst {
+
+  def substitute(subst: Map[TypeVarName, Type], typ: Type): Type =
+    go(freeVars(subst), subst, typ)
+
+  private def go(fv0: Set[TypeVarName], subst0: Map[TypeVarName, Type], typ0: Type): Type =
+    typ0 match {
+      case TVar(name) => subst0.getOrElse(name, typ0)
+      case TTyCon(_) | TBuiltin(_) | TNat(_) => typ0
+      case TApp(t1, t2) => TApp(go(fv0, subst0, t1), go(fv0, subst0, t2))
+      case TForall((v0, k), t) =>
+        if (fv0.contains(v0)) {
+          val v1 = freshTypeVarName(fv0 + v0)
```

Comment:
`fv0.contains(v0)` from above, so:

```suggestion
          val v1 = freshTypeVarName(fv0)
```